### PR TITLE
ViewExtension Framework

### PR DESCRIPTION
**Pull request by aosyatnik**
### Purpose

Link to YouTrack:
[MAGN-7719](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-7719) UI to submit a workspace to the Flood persistence service, so it can be accessed as a customizer

Link to PR with Extension https://github.com/DynamoDS/Dynamo/pull/4670

This PR is a kind of skeleton for real implementation. For me it's better to divide implementation is several parts.

I tried to create `ViewExtension` similar to `Extension`. I still insist on changing architectural design. There are a lot of copied code, e.g. XML loading.

### Impact assessment

- [ ] Changes the way 3rd party nodes work (Add ikeough to the reviewer list)
- [ ] Includes migrations (Add ikeough to the reviewer list)
- [ ] Changes a public interface (Add lukechurch to the reviewer list)
- [x] Affects which components of the system depends on which other components (Add pboyer to the reviewer list)
- [x] Introduces new components (Add lukechurch to the reviewer list)
- [ ] Has a performance impact (Add keyu to the reviewer list)
- [ ] Changes the installer (Add sharadkjaiswal to the reviewer list)

#### References

* [Link to architectural design](https://github.com/DynamoDS/Dynamo/pull/4472)
* [Link to UI design](https://drive.google.com/open?id=0B1DIxgnuFPMlOG5SVVMwVnpkdFE&authuser=1)

#### Testing strategy

Leave it for now. Test project is created.

### Review staging

- [x] Architectural Design Review
- [ ] UI Review

### Reviewers

@pboyer 
@lukechurch

### CODE REVIEW

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 1]
```diff
@@ -113,6 +117,16 @@ public DynamoView(DynamoViewModel dynamoViewModel)
...
             if (dynamoViewModel.Model.AuthenticationManager.HasAuthProvider)
                 dynamoViewModel.Model.AuthenticationManager.AuthProvider.RequestLogin += loginService.ShowLogin;
+
+            var viewExtensions = viewExtensionManager.ExtensionLoader.LoadDirectory(dynamoViewModel.Model.PathManager.ViewExtensionsDirectory);
+            viewExtensionManager.MessageLogged += LogMessage;
+
+            foreach (var ext in viewExtensions)
+            {
+                ext.Startup(null);
+                viewExtensionManager.Add(ext);
```
Comment (pboyer) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs]: We should surround any invocation of an method on an `Extension` with a `try catch`

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 2]
```diff
@@ -420,6 +434,12 @@ private void DynamoView_Loaded(object sender, EventArgs e)
             dynamoViewModel.BeginCommandPlayback(this);
 
             watchSettingsControl.DataContext = background_preview;
+
+            foreach (var ext in viewExtensionManager.ViewExtensions)
+            {
+                ext.Loaded(null);
```
Comment (pboyer) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs]: Surround with `try catch`

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 3]
```diff
@@ -888,6 +908,11 @@ private void WindowClosed(object sender, EventArgs e)
 
             //SHOW or HIDE GALLERY
             dynamoViewModel.RequestShowHideGallery -= DynamoViewModelRequestShowHideGallery;
+
+            foreach (var ext in viewExtensionManager.ViewExtensions)
+            {
+                ext.Shutdown();
```
Comment (pboyer) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs]: Surround with `try catch`

File: src/DynamoPublish/FloodPersistenceServiceExtension.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
+﻿using Dynamo.Interfaces;
+using Dynamo.Wpf.Extensions;
+using System;
+
+namespace Dynamo.Publish
+{
+    public class FloodPersistenceServiceExtension : IViewExtension, ILogSource
```
Comment (pboyer) [inline src/DynamoPublish/FloodPersistenceServiceExtension.cs]: Let's call this `Dynamo Publish` extension.

File: src/DynamoPublishTests/Class1.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace DynamoPublishTests
+{
+    public class Class1
```
Comment (pboyer) [inline src/DynamoPublishTests/Class1.cs]: It's fine that this is empty for now, but let's leave a comment making it clear why it's empty for now.

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 6]
```diff
@@ -113,6 +117,16 @@ public DynamoView(DynamoViewModel dynamoViewModel)
             loginService = new LoginService(this, new System.Windows.Forms.WindowsFormsSynchronizationContext());
             if (dynamoViewModel.Model.AuthenticationManager.HasAuthProvider)
                 dynamoViewModel.Model.AuthenticationManager.AuthProvider.RequestLogin += loginService.ShowLogin;
+
+            var viewExtensions = viewExtensionManager.ExtensionLoader.LoadDirectory(dynamoViewModel.Model.PathManager.ViewExtensionsDirectory);
+            viewExtensionManager.MessageLogged += LogMessage;
```
Comment (pboyer) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs]: Where do we unsubscribe this event?

**Comment by pboyer** [comment]
@aosyatnik,

This looks good to me, except for some minor comments.  Can you fix the code so it is mergeable?

I should clarify why the design is as it is, but I should clarify that I had two intentions here:

1.  Give the implementor control over the Extension lifetime - critical to remove the package manager from core.
2.  Type safety - The code should contain an absolute minimum of casts.  
3.  Minimize reflection - Reflection kills type safety #2 and is painful to maintain.  This is also a reason to avoid inspecting user code for `IMenuItem` implementations.
4.  Obtain buy-in from the rest of the team.  That's why we went through a painful review process here:  https://github.com/DynamoDS/Dynamo/pull/4472

Given that:

1.  I could've used generics.  But that would've resulted in quite a mess, see this discussion:  https://github.com/DynamoDS/Dynamo/pull/4472#discussion_r30554627  They also intertwine the Extension and ViewExtension very much, which is confusing for implementors.  
2.  There really isn't a lot of duplicate code.  The ExtensionLoader and ExtensionManager's are tiny pieces of code, and I fully expect that they will diverge as users have new needs for these components.  By making separate implementations, we've allowed this to happen which makes the code more maintainable and valuable over time.

**Comment by aosyatnik** [comment]
Hi, @pboyer . I've added try-catch statements. I had some problems with adding my project to 2013 solution. If you think there are a lot of commits, I can create new PR with 1 commit. Sorry about that.

**Comment by pboyer** [comment]
Thanks @aosyatnik!  I'm looking forward to seeing the follow-up PR's.

**Pull request merged**